rir/src/compiler/pir/type.h

```diff
@@ -345,6 +346,36 @@ struct PirType {
         return t_.r.includes(o.t_.r);
     }
 
+    void addAssumptions(Assumptions& given, int idx) const {
+        if (!maybeObj())
+            given.setNotObj(idx);
+        if (given.isEager(idx) && given.isNotObj(idx) && isScalar()) {
+            assert(isRType());
+            if (t_.r == RTypeSet(RType::real))
+                given.setSimpleReal(idx);
+            if (t_.r == RTypeSet(RType::integer))
+                given.setSimpleInt(idx);
+        }
+    }
+
+    void loadAssumptions(const Assumptions& given, int idx) {
```

Comment:
this seems at least a more or less a reusable thing. But I don't particularly like that it's here. Does it make sense to tightly couple assumptions and types? would it be better to keep those two things unrelated?